File: CortanaVCDs/cs/adventureworks/viewmodel/tripviewmodel.cs
//*********************************************************
//
// Copyright (c) Microsoft. All rights reserved.
// This code is licensed under the MIT License (MIT).
// THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
//
//*********************************************************

using AdventureWorks.Common;
using AdventureWorks.Model;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Input;
using Windows.Media.SpeechRecognition;
using Windows.ApplicationModel.VoiceCommands;

namespace AdventureWorks.ViewModel
{
    /// <summary>
    /// View Model associated with TripDetails.xaml. Provides access to an individual Trip,
    /// and Commands for saving new, updating existing, and deleting trips.
    /// Is able to both create brand new trips, and edit existing trips, hiding buttons that
    /// should not be accessible in some cases.
    /// </summary>
    public class TripViewModel : ViewModelBase
    {
        private ICommand saveTripCommand;
        private ICommand deleteTripCommand;
        private Trip trip;
        private TripStore store;

        private bool showDestinationValidation = false;
        private string destinationValidationError;

        private bool showDelete = false;

        /// <summary>
        /// The Trip this view model represents.
        /// </summary>
        public Trip Trip
        {
            get
            {
                return trip;
            }
            set
            {
                trip = value;
                NotifyPropertyChanged("Trip");
            }
        }

        /// <summary>
        /// We require that a destination be set to a non-empty string. If the user
        /// attempts to save without one, this will be set to an explanatory validation
        /// prompt to be rendered in the UI.
        /// </summary>
        public String DestinationValidationError
        {
            get
            {
                return destinationValidationError;
            }

            private set
            {
                destinationValidationError = value;
                NotifyPropertyChanged("DestinationValidationError");
            }
        }

        /// <summary>
        /// Controls whether the TextBlock that contains the destination validation string
        /// is visible. Combined with the VisibilityConverter, can show or collapse elements.
        /// </summary>
        public bool ShowDestinationValidation
        {
            get
            {
                return showDestinationValidation;
            }
            private set
            {
                showDestinationValidation = value;
                NotifyPropertyChanged("ShowDestinationValidation");
            }
        }

        /// <summary>
        /// Controls whether the Button that deletes trips is shown. If creating a new trip,
        /// this is false. Otherwise, true.
        /// </summary>
        public bool ShowDelete
        {
            get
            {
                return showDelete;
            }
            private set
            {
                showDelete = value;
                NotifyPropertyChanged("ShowDelete");
            }
        }

        /// <summary>
        /// Bound to the save button, provides a command to invoke when pressed.
        /// </summary>
        public ICommand SaveTripCommand
        {
            get
            {
                return saveTripCommand;
            }
        }

        /// <summary>
        /// Load a trip fomr the store, and set up the view per a normal ShowTrip command.
        /// </summary>
        /// <param name="destination"></param>
        internal void LoadTripFromStore(string destination)
        {
            Trip t = store.Trips.Where(p => p.Destination == destination).FirstOrDefault();
            if (t != null)
            {
                this.ShowTrip(t);
            }
            else
            {
                // Redirect back to the main page, and pass along an error message to show
                App.NavigationService.Navigate<View.TripListView>(
                    string.Format(
                        "Sorry, couldn't find a trip with Destination {0}",
                        destination));
            }
        }


        /// <summary>
        /// Bound to the Delete button, provides a command to invoke when pressed.
        /// </summary>
        public ICommand DeleteTripCommand
        {
            get
            {
                return deleteTripCommand;
            }
        }

        /// <summary>
        /// Construct Trip ViewModel, providing the store to persist trips. 
        /// Creates the RelayCommands to be bound to various buttons in the UI.
        /// </summary>
        /// <param name="store">The persistent store</param>
        public TripViewModel(TripStore store)
        {
            trip = new Trip();
            saveTripCommand = new RelayCommand(new Action(SaveTrip));
            deleteTripCommand = new RelayCommand(new Action(DeleteTrip));
            this.store = store;
        }

        /// <summary>
        /// removes a trip from the store, and returns to the trip list.
        /// </summary>
        private async void DeleteTrip()
        {
            await store.DeleteTrip(Trip);
            await UpdateDestinationPhraseList();
            App.NavigationService.GoBack();
        }

        /// <summary>
        /// Whenever a trip is modified, we trigger an update of the voice command Phrase list. This allows
        /// voice commands such as "Adventure Works Show trip to {destination} to be up to date with saved
        /// Trips.
        /// </summary>
        public async Task UpdateDestinationPhraseList()
        {
            try
            {
                // Update the destination phrase list, so that Cortana voice commands can use destinations added by users.
                // When saving a trip, the UI navigates automatically back to this page, so the phrase list will be
                // updated automatically.
                VoiceCommandDefinition commandDefinitionsEnUs;

                if (VoiceCommandDefinitionManager.InstalledCommandDefinitions.TryGetValue("AdventureWorksCommandSet_en-us", out commandDefinitionsEnUs))
                {
                    List<string> destinations = new List<string>();
                    foreach (Model.Trip t in store.Trips)
                    {
                        destinations.Add(t.Destination);
                    }

                    await commandDefinitionsEnUs.SetPhraseListAsync("destination", destinations);
                }
            }
            catch (Exception ex)
            {
                System.Diagnostics.Debug.WriteLine("Updating Phrase list for VCDs: " + ex.ToString());
            }
        }


        /// <summary>
        /// Performs validation on the destination to ensure it's not empty, then 
        /// saves a trip to the store. If the destination isn't valid, shows a validation
        /// error.
        /// </summary>
        private async void SaveTrip()
        {
            ShowDestinationValidation = false;
            bool valid = true;

            if (String.IsNullOrEmpty(Trip.Destination))
            {
                valid = false;
                ShowDestinationValidation = true;
                DestinationValidationError = "Destination cannot be blank";
            }
            else
            {
                Trip.Destination = Trip.Destination.Trim();
            }

            if (valid)
            {
                await store.SaveTrip(this.Trip);
                await UpdateDestinationPhraseList();
                App.NavigationService.GoBack();
            }
        }

        /// <summary>
        /// Sets up the view model to show an existing trip.
        /// </summary>
        /// <param name="trip"></param>
        internal void ShowTrip(Trip trip)
        {
            showDelete = true;
            Trip = trip;
        }

        /// <summary>
        /// Sets up the view model to create a new trip.
        /// </summary>
        internal void NewTrip()
        {
            showDelete = false;
            Trip = new Trip();
        }
    }
}

File: Samples/CortanaVoiceCommand/cpp/AdventureWorks/ViewModel/TripViewModel.cpp
//*********************************************************
//
// Copyright (c) Microsoft. All rights reserved.
// This code is licensed under the MIT License (MIT).
// THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
//
//*********************************************************
#include "pch.h"
#include "TripViewModel.h"
#include "Common\RelayCommand.h"
#include "View\TripListView.xaml.h"

using namespace AdventureWorks;

using namespace Concurrency;
using namespace AdventureWorks_Shared;
using namespace Platform;
using namespace Platform::Collections;
using namespace Windows::ApplicationModel::VoiceCommands;
using namespace Windows::Globalization;
using namespace Windows::UI::Xaml;
using namespace Windows::UI::Xaml::Controls;
using namespace Windows::UI::Xaml::Input;

/// <summary> Handle initialization of command data binding </summary>
TripViewModel::TripViewModel(TripStore^ store)
{
    this->store = store;
    saveTripCommand = ref new RelayCommand(
        [this](Object^ parameter) {return true; },
        [this](Object^ parameter) {this->SaveTrip(); });

    deleteTripCommand = ref new RelayCommand(
        [this](Object^ parameter) {return true; },
        [this](Object^ parameter) {this->DeleteTrip(); });

}

AdventureWorks_Shared::Trip^ TripViewModel::Trip::get()
{
    return this->trip;
}

void TripViewModel::Trip::set(AdventureWorks_Shared::Trip^ value)
{
    this->trip = value;
    NotifyPropertyChanged("Trip");
}

String^ TripViewModel::DestinationValidationError::get()
{
    return this->destinationValidationError;
}

void TripViewModel::setDestinationValidation(String^ value)
{
    this->destinationValidationError = value;
    NotifyPropertyChanged("DestinationValidationError");
}

bool TripViewModel::ShowDestinationValidation::get()
{
    return this->showDestinationValidation;
}

void TripViewModel::setShowDestinationValidation(bool value)
{
    this->showDestinationValidation = value;
    NotifyPropertyChanged("ShowDestinationValidation");
}

bool TripViewModel::ShowDelete::get()
{
    return this->showDelete;
}

void TripViewModel::setShowDelete(bool value)
{
    this->showDelete = value;
    NotifyPropertyChanged("ShowDelete");
}

ICommand^ TripViewModel::SaveTripCommand::get()
{
    return this->saveTripCommand;
}

ICommand^ TripViewModel::DeleteTripCommand::get()
{
    return this->deleteTripCommand;
}

/// <summary> 
/// Load trip data (if required), and then search the store for the specified trip.
/// Does not disambiguate, just displays the first match, for simplicity in this sample.
/// Navigates back to the main page if it can't be found (covers the case where a trip was removed
/// but the phrase list was not updated.)
/// <summary>
/// <param name="destination">The destination to search for</param>
void TripViewModel::LoadTripFromStore(String^ destination)
{
    create_task(store->LoadTripsAsync()).then([this, destination]() {

        for (unsigned int i = 0; i < store->Trips->Size; i++)
        {
            AdventureWorks_Shared::Trip^ t = store->Trips->GetAt(i);
            if (t->Destination == destination)
            {
                this->ShowTrip(t);
                return;
            }
        }

        String^ errorMsg = ref new String(L"Sorry, couldn't find a trip with Destination ") + destination;
        auto rootFrame = dynamic_cast<Frame^>(Window::Current->Content);
        rootFrame->Navigate(TripListView::typeid, errorMsg);
    });
}

/// <summary>Set the data-bound trip (and show delete button)</summary>
/// <param name="trip">The trip to show</param>
void TripViewModel::ShowTrip(AdventureWorks_Shared::Trip^ trip)
{
    setShowDelete(true);
    this->Trip = trip;
}

/// <summary>Create a new trip, hide the delete button.</summary>
void TripViewModel::NewTrip()
{
    setShowDelete(false);
    Trip = ref new AdventureWorks_Shared::Trip();
}

///<summary> Delete the currently displayed trip from the trip store,
/// trigger update of phrase list.</summary>
void TripViewModel::DeleteTrip()
{
    store->DeleteTripAsync(this->Trip);

    this->UpdateDestinationPhraseList();
    auto rootFrame = dynamic_cast<Frame^>(Window::Current->Content);
    rootFrame->Navigate(TripListView::typeid);

}

/// <summary> Update the phrase list, issued on initial startup of the app, and 
/// whenever a trip is modified and saved.</summary>
void TripViewModel::UpdateDestinationPhraseList()
{
    VoiceCommandDefinition^ commandDefinitions;
	auto languages = Windows::Globalization::ApplicationLanguages::Languages;
	String^ countryCode = L"en-US";
	if (languages->Size > 0)
	{
		// Languages is ordered by the preference order that's specified in settings.
		countryCode = languages->GetAt(0);
	}

    if (VoiceCommandDefinitionManager::InstalledCommandDefinitions->HasKey(L"AdventureWorksCommandSet_" + countryCode))
    {
        commandDefinitions = VoiceCommandDefinitionManager::InstalledCommandDefinitions->Lookup(L"AdventureWorksCommandSet_" + countryCode);

        Vector<String^>^ destinations = ref new Vector<String^>();
        int size = this->store->Trips->Size;
        for (int i = 0; i < size; i++)
        {
            AdventureWorks_Shared::Trip^ t = this->store->Trips->GetAt(i);
            String^ destination = t->Destination;
            unsigned int unused;
            if (!destinations->IndexOf(destination, &unused))
            {
                destinations->Append(destination);
            }
        }

        create_task(commandDefinitions->SetPhraseListAsync(L"destination", destinations))
            .then([](Concurrency::task<void> t) {
            try
            {
                t.get();
            }
            catch (Platform::Exception^ ex)
            {
                OutputDebugStringW(L"Failed to install phrase set into AdventureWorksCommandSet_en-us\n");
                OutputDebugStringW(ex->Message->Data());
            }
        });
    }
}

/// <summary> Save a trip to the store, trigger an update of the phrase list, 
/// then navigate back to the main page </summary>
void TripViewModel::SaveTrip()
{
    setShowDestinationValidation(false);
    bool valid = true;

    if ((Trip->Destination->IsEmpty()))
    {
        valid = false;
        setShowDestinationValidation(true);
        setDestinationValidation(L"Destination cannot be blank");
    }

    if (valid)
    {
        create_task(store->SaveTripAsync(this->Trip), task_continuation_context::use_current()).then(
            [this](task<void> saveTripTask)
        {
            saveTripTask.get();
            this->UpdateDestinationPhraseList();
            auto rootFrame = dynamic_cast<Frame^>(Window::Current->Content);
            rootFrame->Navigate(TripListView::typeid);
        });
    }
}
